coap-gateway/service/signOff.go

```diff
@@ -109,26 +109,34 @@ func signOffHandler(req *mux.Message, client *Client) {
 
 	jwtClaims, err := client.ValidateToken(req.Context, signOffData.accessToken)
 	if err != nil {
-		logErrorAndCloseClient(fmt.Errorf("cannot handle sign off: %v", err), coapCodes.InternalServerError)
+		logErrorAndCloseClient(fmt.Errorf("cannot handle sign off: %v", err), coapCodes.Unauthorized)
 		return
 	}
 
-	if err := jwtClaims.validateOwnerClaim(client.server.config.Clients.AuthServer.OwnerClaim, signOffData.userID); err != nil {
-		logErrorAndCloseClient(fmt.Errorf("cannot handle sign off: %v", err), coapCodes.InternalServerError)
+	err = client.server.VerifyDeviceID(client.tlsDeviceID, jwtClaims)
+	if err != nil {
+		logErrorAndCloseClient(fmt.Errorf("cannot handle sign off: %w", err), coapCodes.Unauthorized)
+		return
+	}
+
+	if err := jwtClaims.ValidateOwnerClaim(client.server.config.Clients.AuthServer.OwnerClaim, signOffData.userID); err != nil {
+		logErrorAndCloseClient(fmt.Errorf("cannot handle sign off: %v", err), coapCodes.Unauthorized)
 		return
 	}
 
+	deviceID := client.server.GetDeviceID(jwtClaims, client.tlsDeviceID, signOffData.deviceID)
+
 	ctx = kitNetGrpc.CtxWithToken(ctx, signOffData.accessToken)
-	deviceIds := []string{signOffData.deviceID}
+	deviceIds := []string{deviceID}
 	respRA, err := client.server.raClient.DeleteDevices(ctx, &commands.DeleteDevicesRequest{
 		DeviceIds: deviceIds,
 	})
 	if err != nil {
-		logErrorAndCloseClient(fmt.Errorf("cannot handle sign off: %w", err), coapCodes.InternalServerError)
+		logErrorAndCloseClient(fmt.Errorf("cannot handle sign off: %w", err), coapCodes.Unauthorized)
```

Comment:
I think `DeleteDevices` can also fail because of some mongodb error, returning Unathorized doesn't seem correct in that case. Though `raClient.DeleteDevices` always returns `codes.Internal`, we should updated to `Unauthenticated` when user doesn't have access and then translate the code here to correct coap code.

Reply:
Fixed